doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,715 @@
+    For example ``PyArray_Descr->f->nonzero`` or ``PyArray_Descr->f->copyswapn``,
+    may instead receive a dummy array object with only some fields (mainly the
+    dtype), being valid.
+    At least in some code paths, a similar mechanism is already used.
+
+  * The ``scalarkind`` slot and registration of scalar casting will be
+     removed/ignored without replacement.
+     It currently allows partial value-based casting.
+     The ``PyArray_ScalarKind`` function will continue to work for builtin types,
+     but will not be used internally and be deprecated.
+
+   * Currently user dtypes are defined as instances of ``np.dtype``.
+     The creation works by the user providing a prototype instance.
+     NumPy will need to modify at least the type during registration.
+     This has no effect for either ``rational`` or ``quaternion`` and mutation
+     of the structure seems unlikely after registration.
+
+Since there is a fairly large API surface concerning datatypes, further changes
+or the limitation certain function to currently existing datatypes is
+likely to occur.
+For example functions which use the type number as input
+should be replaced with functions taking DType classes instead.
+Although public, large parts of this C-API seem to be used rarely,
+possibly never, by downstream projects.
+
+
+
+Detailed Description
+--------------------
+
+This section details the design decisions covered by this NEP.
+The subsections correspond to the list of design choices presented
+in the Scope section.
+
+Datatypes as Python Classes (1)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The current NumPy datatypes are not full scale python classes.
+They are instead (prototype) instances of a single ``np.dtype`` class.
+Changing this means that any special handling, e.g. for ``datetime``
+can be moved to the Datetime DType class instead, away from monolithic general
+code (e.g. current ``PyArray_AdjustFlexibleDType``).
+
+The main consequence of this change with respect to the API is that
+special methods move from the dtype instances to methods on the new DType class.
+This is the typical design pattern used in Python.
+Organizing these methods and information in a more Pythonic way provides a
+solid foundation for refining and extending the API in the future.
+The current API cannot be extended due to how it is exposed publically.
+This means for example that the methods currently stored in ``ArrFunctions``
+on each datatype (see NEP 40) will be defined differently in the future and
+deprecated in the long run.
+
+The most prominent visible side effect of this will be that
+``type(np.dtype(np.float64))`` will not be ``np.dtype`` anymore.
+Instead it will be a subclass of ``np.dtype`` meaning that
+``isinstance(np.dtype(np.float64), np.dtype)`` will remain true.
+This will also add the ability to use ``isinstance(dtype, np.dtype[float64])``
+thus removing the need to use ``dtype.kind``, ``dtype.char``, or ``dtype.type``
+to do this check.
```

Comment:
In my opinion no one should have been doing the check this way anyway - the failsafe way to spell that has always been simply `dt == float64`, or `np.dtype(dt) == float64` if you don't know for sure whether you have a dtype or not.

Reply:
True, we do use the `kind` check for `np.issubdtype` checks in some places I think... But that is, to some degree, for later.